sympy/matrices/eigen.py

```diff
@@ -1357,6 +1362,142 @@ def pick_vec(small_basis, big_basis):
 
     return restore_floats2(basis_mat, jordan_mat)
 
+def jordan_form_rational_matrix(M, calc_transform):
+    """
+    Return (P, J), where J is the Jordan form and P satisfies M = P J P^{-1},
+    if the matrix is rational. If calc_transform is False, return only J.
+    """
+
+    # cache calculations for some speedup
+    char_mat_cache = {}
+
+    def char_mat(val, pow):
+        """Cache computations of ``(M - val*I)**pow`` for quick
+        retrieval"""
+
+        if (val, pow) in char_mat_cache:
+            return char_mat_cache[(val, pow)]
+
+        else:
+            eigen_scalar = a if a is not None else val
+            char_mat_cache[(val, pow)] = (M - eigen_scalar*M.eye(M.rows)).to_DM(extension=True).pow(pow)
+        return char_mat_cache[(val, pow)]
+
+    nullspace_cache = {}
+
+    def nullspace_matrix(eig, pow):
+        if a is not None:
+            if pow not in nullspace_cache:
+                mat = char_mat(eig, pow)
+                eig_mat = mat.nullspace().to_Matrix()
+                nullspace_cache[pow] = eig_mat
+            else:
+                eig_mat = nullspace_cache[pow]
+            eig_mat = eig_mat.subs(a, eig)
+        else:
+            mat = char_mat(eig, pow)
+            eig_mat = mat.nullspace().to_Matrix()
+        return eig_mat.T
+
+    # helper functions
+    def nullity_chain(val, algebraic_multiplicity):
+        """Calculate the sequence  [0, nullity(E), nullity(E**2), ...]
+        until it is constant where ``E = M - val*I``"""
+
+        ret     = [0]
+
+        nullity = len(nullspace_to_list(nullspace_matrix(val, 1)))
+        i       = 2
+
+        while nullity != ret[-1]:
+            ret.append(nullity)
+
+            if nullity == algebraic_multiplicity:
+                break
+            nullity = len(nullspace_to_list(nullspace_matrix(val, i)))
+            i       += 1
+
+        return ret
+
+    def nullspace_to_list(mat):
+        return [mat.col(i) for i in range(mat.cols)]
+
+    a = None # Algebraic number
+    char_eq = M.charpoly().as_expr()
+    eigenvalues = dict(Counter(all_roots(char_eq)))
+    for i in eigenvalues.keys():
+         if not isinstance(i,Integer):
+            a = AlgebraicNumber(i, alias='a')
+            break
```

Comment:
I don't see how you can have only a single AlgebraicNumber. If there is one irrational eigenvalue then there will always be more than one because it will be the root of a factor that is at least degree 2.